bluetooth: host: add checks for connection types

kozdra opened the pull request:
Fail gracefully if an HCI event of one type arrives for a handle of a different connection type.  The requested types are currently based on what fields are used, not on the usage context, in order to keep every correct use so far still working.

A warning is logged if the connection identified by the handle does not match the requested connection type.

kozdra committed:
bluetooth: host: add checks for connection types

Fail gracefully if an HCI event of one type arrives for a handle of a
different connection type.  The requested types are currently based on
what fields are used, not on the usage context, in order to keep every
correct use so far still working.

A warning is logged if the connection identified by the handle does not
match the requested connection type.

Signed-off-by: Arkadiusz Kozdra <akozdra@antmicro.com>

Thalley commented:
> Fail gracefully if an HCI event of one type arrives for a handle of a different connection type. The requested types are currently based on what fields are used, not on the usage context, in order to keep every correct use so far still working.
> 
> A warning is logged if the connection identified by the handle does not match the requested connection type.

This seems to guard against a faulty controller, right? Has this been an actual issue, or just defensive programming?

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1205,28 +1207,37 @@ void bt_conn_set_state(struct bt_conn *conn, bt_conn_state_t state)
 	conn = conn_lookup_handle(acl_conns, ARRAY_SIZE(acl_conns), handle);
 	if (conn) {
-		return conn;
+		if (type & conn->type) {
+			return conn;
+		}
+		LOG_WRN("incompatible handle %u", handle);
```
There's a reference leak here, i.e. you need to do `bt_conn_unref()` if you don't intend to return the conn pointer.

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1205,28 +1207,37 @@ void bt_conn_set_state(struct bt_conn *conn, bt_conn_state_t state)
 	conn = conn_lookup_handle(iso_conns, ARRAY_SIZE(iso_conns), handle);
 	if (conn) {
-		return conn;
+		if (type & conn->type) {
+			return conn;
+		}
+		LOG_WRN("incompatible handle %u", handle);
```
This part should at most be an assert, I think, since `iso_conns` should only contain BT_CONN_TYPE_ISO objects.

kozdra replied:
We search through the list regardless of the passed type, though.  We could make it more like `bt_conn_foreach` iterating only over the correct lists (or even use the foreach function, but all my ideas involve `setjmp`, which is not quite pretty either), but then it would miss the warnings and just return the NULL.

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1205,28 +1207,37 @@ void bt_conn_set_state(struct bt_conn *conn, bt_conn_state_t state)
+			return conn;
+		}
+		LOG_WRN("incompatible handle %u", handle);
 	}
 #endif
 
- #if defined(CONFIG_BT_BREDR)
+#if defined(CONFIG_BT_BREDR)
 	conn = conn_lookup_handle(sco_conns, ARRAY_SIZE(sco_conns), handle);
 	if (conn) {
-		return conn;
+		if (type & conn->type) {
+			return conn;
+		}
+		LOG_WRN("incompatible handle %u", handle);
```
Same here. `sco_conns` should only contain `BT_CONN_TYPE_SCO` objects

jhedberg commented on subsys/bluetooth/host/direction.c:
```diff
@@ -680,7 +680,7 @@ int hci_df_prepare_connection_iq_report(struct net_buf *buf,
 	evt = net_buf_pull_mem(buf, sizeof(*evt));
 
-	conn = bt_conn_lookup_handle(sys_le16_to_cpu(evt->conn_handle));
+	conn = bt_conn_lookup_handle(sys_le16_to_cpu(evt->conn_handle), BT_CONN_TYPE_ALL);
```
I'm pretty sure Direction Finding connections are always normal LE connections.

jhedberg commented on subsys/bluetooth/host/direction.c:
```diff
@@ -732,7 +732,7 @@ int hci_df_vs_prepare_connection_iq_report(struct net_buf *buf,
 	evt = net_buf_pull_mem(buf, sizeof(*evt));
 
-	conn = bt_conn_lookup_handle(sys_le16_to_cpu(evt->conn_handle));
+	conn = bt_conn_lookup_handle(sys_le16_to_cpu(evt->conn_handle), BT_CONN_TYPE_ALL);
```
Same here, i.e. DF connections should always be of type LE, I believe.

jhedberg commented on subsys/bluetooth/host/direction.c:
```diff
@@ -857,7 +857,7 @@ int hci_df_prepare_conn_cte_req_failed(struct net_buf *buf,
 	evt = net_buf_pull_mem(buf, sizeof(*evt));
 
-	conn = bt_conn_lookup_handle(sys_le16_to_cpu(evt->conn_handle));
+	conn = bt_conn_lookup_handle(sys_le16_to_cpu(evt->conn_handle), BT_CONN_TYPE_ALL);
```
And here.

jhedberg commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1229,7 +1229,7 @@ void hci_le_cis_req(struct net_buf *buf)
 
 	/* Lookup existing connection with same handle */
-	iso = bt_conn_lookup_handle(cis_handle);
+	iso = bt_conn_lookup_handle(cis_handle, BT_CONN_TYPE_ALL);
```
Should probably be `BT_CONN_TYPE_ISO`?

jhedberg commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1238,7 +1238,7 @@ void hci_le_cis_req(struct net_buf *buf)
 
 	/* Lookup ACL connection to attach */
-	acl = bt_conn_lookup_handle(acl_handle);
+	acl = bt_conn_lookup_handle(acl_handle, BT_CONN_TYPE_ALL);
```
Should probably be `BT_CONN_TYPE_LE`. @Thalley can you confirm?

Thalley replied:
Yes, a "parent" connection to a CIS is always a `BT_CONN_TYPE_LE`

jhedberg commented on subsys/bluetooth/host/scan.c:
```diff
@@ -1201,7 +1201,8 @@ static void bt_hci_le_past_received_common(struct net_buf *buf)
 	sync_info.conn = bt_conn_lookup_handle(
-				sys_le16_to_cpu(evt->conn_handle));
+				sys_le16_to_cpu(evt->conn_handle),
+				BT_CONN_TYPE_ALL);
```
I suspect this should be `BT_CONN_TYPE_LE`

jhedberg commented on subsys/bluetooth/host/ssp.c:
```diff
@@ -779,7 +779,7 @@ void bt_hci_auth_complete(struct net_buf *buf)
 	LOG_DBG("status 0x%02x, handle %u", evt->status, handle);
 
-	conn = bt_conn_lookup_handle(handle);
+	conn = bt_conn_lookup_handle(handle, BT_CONN_TYPE_ALL);
```
Everything in `ssp.c` is specific to BR/EDR, so this should be `BT_CONN_TYPE_BR`

jhedberg reviewed (changes requested):
Thanks for submitting the PR. Please see my inline comments for needed improvements.

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1205,28 +1207,40 @@ void bt_conn_set_state(struct bt_conn *conn, bt_conn_state_t state)
 	conn = conn_lookup_handle(sco_conns, ARRAY_SIZE(sco_conns), handle);
 	if (conn) {
-		return conn;
+		if (type & conn->type) {
+			return conn;
+		}
+		LOG_WRN("incompatible handle %u", handle);
```
Instead of having this check 3 times, suggest to refactor the function to only have a single `return` statement with a single check for type.

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1205,28 +1207,40 @@ void bt_conn_set_state(struct bt_conn *conn, bt_conn_state_t state)
 }
 
-struct bt_conn *bt_conn_lookup_handle(uint16_t handle)
+struct bt_conn *bt_conn_lookup_handle(uint16_t handle, int type)
```
`type` should be an `uint8_t`.

jhedberg replied:
> `type` should be an `uint8_t`.

If we're going to fine-tune this part, I think it'd be cleaner to just give a name to the enum and use that instead of generic int types.

kozdra replied:
I pushed my suggested solution into a separate commit so that it can be reviewed easier.

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1205,28 +1207,40 @@ void bt_conn_set_state(struct bt_conn *conn, bt_conn_state_t state)
 }
 
-struct bt_conn *bt_conn_lookup_handle(uint16_t handle)
+struct bt_conn *bt_conn_lookup_handle(uint16_t handle, int type)
```
This feels like the wrong way of doing this IMO. 

`bt_conn_lookup_handle` is a local function to get a local `bt_conn` based on a local handle. By "local" I mean this isn't related to the controller, which is what we seem to want to guard against. 

IMO any caller of this function that want to verify the type, should do that on its own. Ideally this should only be done when looking up a handle provided by the controller, and not everywhere.

jhedberg replied:
> `bt_conn_lookup_handle` is a local function to get a local `bt_conn` based on a local handle. By "local" I mean this isn't related to the controller, which is what we seem to want to guard against.

Could you clarify a bit what you mean by the above? `struct bt_conn` is a "local" host-side abstraction for connections, however conn->handle is always the value we get from the controller and what we pass back to the controller to refer to a connection.

Thalley replied:
> however conn->handle is always the value we get from the controller and what we pass back to the controller to refer to a connection.

I think I need to retract my previous comment. I was under the impression that `bt_conn_lookup_handle` was used in other contexts besides getting a `bt_conn` from a controller event, but it seems that I am incorrect; it seems to be exclusively used for that purpose, in which case it makes sense to do it this way. 

_if_ it was used by the host from other host actions, then I think my comment still applies, but I agree that the way that this is used currently, this is a OK way of doing it. 

My bad for not checking properly before commenting :)

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -390,7 +391,7 @@ struct bt_security_info {
 struct bt_conn_info {
 	/** Connection Type. */
-	uint8_t type;
+	enum bt_conn_type type : 8;
```
```suggestion
	enum bt_conn_type type;
```

Generally I don't think we want to use bitfields in public header files, as architecture support of these may vary.

Thalley replied:
@jhedberg correct me if I am wrong here :)

kozdra replied:
Although I am not strictly encouraging the use of bitfields, I decided to use them (as they *should* keep the old ABI and not inflate the necessary RAM needlessly) because a quick search revealed that Zephyr uses enum bitfields already in two USB-related public facing headers:

https://github.com/zephyrproject-rtos/zephyr/blob/6050a10f8b298fe6994bb47ab9775ba535a604dd/include/zephyr/drivers/usb/udc.h#L44

https://github.com/zephyrproject-rtos/zephyr/blob/6050a10f8b298fe6994bb47ab9775ba535a604dd/include/zephyr/drivers/usb_c/usbc_pd.h#L483

jhedberg replied:
The approach we've taken to compress enums into smaller space (in some few places where we thought it mattered), is to use `__packed` instead of bitfields, like here:
https://github.com/zephyrproject-rtos/zephyr/blob/90355f143f21d9ba985f62ea95d96ff83f6922ef/subsys/bluetooth/host/conn_internal.h#L14-L24

I think that'd be a better choice, since it ensures that the resulting size grows if we end up having more values than can fit a `uint8_t`.

Thalley replied:
`__packed` is indeed a nice way to keep it small while still supporting expansions.

Thalley commented on subsys/bluetooth/host/conn_internal.h:
```diff
@@ -159,7 +159,7 @@ struct acl_data {
 struct bt_conn {
 	uint16_t			handle;
-	uint8_t			type;
+	enum bt_conn_type	type : 8;
```
```suggestion
	enum bt_conn_type	type;
```

Even though this isn't a public header, my comment w.r.t. bitfields in header file still applies. 

Unless it is required to be specific size, suggest to just use the `enum` without the bitfield.

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -196,6 +196,21 @@ struct bt_conn_le_data_len_param {
 				  BT_GAP_DATA_TIME_MAX)
 
+/** Connection Type */
+typedef enum __packed {
```
No reason to `typedef` this :) Suggest to keep as `enum bt_conn_type_t`

kozdra replied:
I thought `__packed` can only be applied on `typedef`s, but if this is wrong, I can make it a simple `enum bt_conn_type` again.

jhedberg replied:
FWIW, I have nothing against a making a typedef for connection types, however I'm also not sure where and how exactly `__packed` can be used when it comes to enums. Maybe it's possible to not have it in the enum definition but only use it when declaring a variable - try it out and see what works.

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -233,7 +248,8 @@ void bt_conn_unref(struct bt_conn *conn);
  * @param data  Data to pass to the callback function.
  */
-void bt_conn_foreach(int type, void (*func)(struct bt_conn *conn, void *data),
+void bt_conn_foreach(bt_conn_type_t type,
+		     void (*func)(struct bt_conn *conn, void *data),
```
I guess this is technically changing a stable API, right @jhedberg ? Should this go through the https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-incompatible-changes ?

kozdra committed:
bluetooth: host: track connection type enum

The enum used for connection types gets named bt_conn_type to guard
against accidental usage of generic integers with relation to it.

The added default case in several switch statements avoids warnings
against unhandled enum values.

Signed-off-by: Arkadiusz Kozdra <akozdra@antmicro.com>

jori-nordic commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -233,7 +248,8 @@ void bt_conn_unref(struct bt_conn *conn);
  * @param data  Data to pass to the callback function.
  */
-void bt_conn_foreach(int type, void (*func)(struct bt_conn *conn, void *data),
+void bt_conn_foreach(enum bt_conn_type type,
```
this should be a new api `bt_conn_foreach_type`

kozdra replied:
I think I disagree that the API changes actually, in my opinion it only adds safety guards against bad ideas.  The 'sane' way to use the function, i.e. calling `bt_conn_foreach(BT_CONN_TYPE_*, callback, pointer)` stays functional.  Supplying the first argument from some intermediate `int` variable is not breaking any usage either, since it does not even print warnings in GCC nor Clang (and `int` is guaranteed to be large enough to hold `enum` values; especially here, there are only 16 possible valid values after all).  The only thing that changes is that it guards against redeclaration with int instead of enum (only a warning when `-Wall` implies `-Wenum-int-mismatch`) and passing in a value from a different enum — and again, this is only a compiler warning (`-Wextra` implies `-Wenum-conversion`), not an error.

There *might be* some wild use cases of course if someone called typeof() on the `bt_conn_foreach` function, or they took address of it, but they will all(?) still work except for maybe some really exotic ABIs passing the `enum` argument differently than promoting it to `int` under the hood. The compiler will regardless show `-Wincompatible-pointer-types` for this case, which is on by default.

If you know any other thing that could break and that I did not think about, or if the project policy mandates making this a new API just to be on the safe side (even if it is in fact compatible), please let me know.

jori-nordic replied:
I'm so sorry, I got confused by the signature and the diff. I thought the `type` param was new. 
Yeah this is fine, sorry about that
